### ISSUE TITLE
no-useless-return clashes with callback-return when return is not referencing the callback

### ISSUE THREAD

**Issue by gajus** [eslint/eslint#7626]
**Tell us about your environment**

* **ESLint Version:**

v3.10.2

* **Node Version:**

v7.1.0

* **npm Version:**

v4.0.2

**What parser (default, Babel-ESLint, etc.) are you using?**

Default.

**Please show your full configuration:**

```json
{
  "rules": {
    "no-useless-return": 2,
    "callback-return": 2
  }
}
```

**What did you do? Please include the actual source code causing the issue.**

```js
// Example A
// Reports "callback-return" (expected)

() => {
  if (typeof cb === 'function') {
    cb();
  }
};

// Example B
// Reports "callback-return" (expected)
() => {
  if (typeof cb === 'function') {
    return cb();
  }
};

// Example C
// Reports "no-useless-return" (***unexpected***)
() => {
  if (typeof cb === 'function') {
    cb();
    
    return;
  }
};
```

Using `no-useless-return` and `callback-return` enforces to use the code style in example B. The problem with example B is that it enables the callback function to return a value.

Using code style shown in example C enables to read easily that function does not return any value.

`no-useless-return` should work with `callback-return` to recognise when a return is preceded by a callback function and not report it as a warning.

**Comment by not-an-aardvark** [comment]
Thanks for the report. It looks like this was also noted in https://github.com/eslint/eslint/pull/7441, but we decided to add no-useless-return as-is anyway.

In my opinion, the best solution to this problem would be to have `callback-return` not require return statements where they are redundant anyway.

**Comment by platinumazure** [comment]
@not-an-aardvark Now that we have code path analysis, I think your suggestion is quite feasible and makes sense. I would be :+1: to augmenting `callback-return` to not require return statements when they are redundant (if that changes due to a code edit, the rule would catch the issue).

**Comment by not-an-aardvark** [comment]
I don't think this would be a semver-major change, since it would report fewer errors by default. So I don't think it would be necessary to put the behavior behind an option.

**Comment by platinumazure** [comment]
@not-an-aardvark I wasn't speaking in terms of semver-major and API compatibility, but rather for users who might not use `no-useless-return` and who might prefer a useless return on top of a callback anyway (i.e., almost for stylistic reasons). If there are users like that right now, then this proposed enhancement would be harmful to them.

**Comment by nzakas** [comment]
Unfortunately, it looks like there wasn't enough interest from the team or community to implement this change. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to be implemented after 90 days tend to never be implemented, and as such, we [close those issues](https://eslint.org/docs/maintainer-guide/issues#when-to-close-an-issue). This doesn't mean the idea isn't interesting or useful, just that it's not something the team can commit to.
